README.md
# HLearn
To train MNIST with GPU change back-end in Backprop.hs to Data.Array.Accelerate.LLVM.PTX
To train MNIST with CPU change back-end in Backprop.hs to Data.Array.Accelerate.LLVM.Native

A directory named "data" must be located in the same directory as the resulting executable after compilation. It must contain the MNIST database as .idx. 
You can download MNIST from http://yann.lecun.com/exdb/mnist/

When compiling for the GPU you must pass the additional flag -threaded to GHC


CHANGELOG.md
# Revision history for HLearn

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/Ops.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE TypeApplications                         #-}
{-# LANGUAGE TypeOperators                         #-}
{-# LANGUAGE ScopedTypeVariables                      #-}
{-# LANGUAGE ViewPatterns #-}
{-# LANGUAGE Rank2Types #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE BangPatterns #-}

module Ops where
import Backprop
import Lift
import qualified Data.Array.Accelerate as A
import Data.Array.Accelerate (Exp(..)
                             ,Arrays(..)
                             ,Acc(..)
                             ,(:.)(..)
                             ,Z (..))
import Data.Array.Accelerate.System.Random.MWC
import qualified Prelude as P 
import Prelude hiding ((++), map, zipWith, (<>))
import Numeric.Backprop
import GHC.TypeLits
import Data.Proxy
import Data.List (foldl')
--------------------------------------------------------------------------------
class RandomArray a where
  sample :: IO (a)
instance (KnownNat n) => RandomArray (SR n) where
  sample = do a <- randomArray uniform (Z:.n) :: IO(A.Vector ℝ)
              return (pack @(SR n) a)
    where n = fromInteger $ natVal (Proxy @n) :: Int
  -- sample = return (run 0)
instance (KnownNat m, KnownNat n) => RandomArray (SL m n) where
  sample = do a <- randomArray uniform (Z:.m:.n) :: IO (A.Matrix ℝ)
              return (pack @(SL m n) a)
    where n = fromInteger $  natVal (Proxy @n) :: Int
          m = fromInteger $  natVal (Proxy @m) :: Int
instance (RandomArray a, RandomArray b) => RandomArray (a,b) where
  sample = do a <- sample
              b <- sample
              return (a,b)
--------------------------------------------------------------------------------

type Rate = Exp ℝ
type Learn p a b = forall z. (Backprop p, Backprop a, Backprop b, Reifies z W)
                 => BVar z p -> BVar z a -> BVar z b


-- Post-composition of learners
(~>) :: (Backprop p,
         Runable p,
         Backprop q,
         Runable q,
         Backprop b,
         Runable b,
         Reifies z W)
     =>  (BVar z p -> BVar z a -> BVar z b)
     ->  (BVar z q -> BVar z b -> BVar z c)
     ->  (BVar z (p ⊗ q) -> BVar z a -> BVar z c)
(~>) i j (p:*:q) = j q . compute . i p

(⊗) :: (Backprop p,
        Runable p,
        Backprop q,
        Runable q,
        Backprop a,
        Runable a,
        Backprop b,
        Runable b,
        Backprop c,
        Runable c,
        Backprop d,
        Runable d)
    => (Learn p a c)
    -> (Learn q b d)
    -> (Learn (p ⊗ q) (a ⊗ b) (c ⊗ d))
(⊗) i j (p:*:q) (a:*:b) = (i p a):*:(j q b)

type Error a b = forall z. (Backprop a, Backprop b, Reifies z W)
               => BVar z a -> BVar z a -> BVar z b
errorGrad :: (Backprop p,
              Backprop a,
              Backprop b,
              Runable c,
              Backprop c)
          => (forall z. Reifies z W => BVar z p -> BVar z a -> BVar z b)
          -> (forall s. Reifies s W => BVar s b -> BVar s b -> BVar s c)
          -> a
          -> b
          -> p
          -> p
errorGrad i e a b = gradBP $ \p -> e (i p $ auto a) (auto b)



logistic :: Floating a => a -> a
logistic x = 1 / (1 + exp (-x))

cmpr :: forall z.
        (Reifies z W)
     => (Exp ℝ -> Exp ℝ -> Exp Bool) 
     -> BVar z (Exp ℝ)
     -> BVar z (Exp ℝ)
     -> BVar z (Exp ℝ)
     
cmpr f = liftOp2 . op2 $ \x y -> (A.cond (f x y) x y,
                               (\d -> A.unlift $ A.cond (f x y)
                                      (A.T2 d 0)
                                      (A.T2 0 d)
                               ))
                              
  


konstR :: forall m z.
          (KnownNat m,
           Reifies z W) 
       => BVar z (R 1)
       -> BVar z (R m)
konstR = liftOp1 . op1 $ \(unpack -> x) -> let val = x A.!! 0
                                               m = fromInteger $ natVal (Proxy @m)
                                           in (pack $ A.fill (A.constant $ Z:.m) val,
                                               evalBP sumR)

replicateL :: forall n m k.
              (KnownNat n,
               KnownNat m)
           => (R n)
           -> (L m n)
replicateL (unpack -> v) = let m = fromInteger $ natVal (Proxy @m) :: Int
                           in pack $ A.replicate (A.constant $ Z:.m:.A.All) v
 
konstL :: forall n m z.
          (KnownNat n,
           KnownNat m,
           Reifies z W) 
       => BVar z (R m)
       -> BVar z (L m n)
konstL = liftOp1 . op1 $ \(unpack -> x) -> let n = fromInteger $ natVal (Proxy @n) :: Int
                                           in (pack $ A.replicate (A.constant $ Z:.A.All:.n) x,
                                               evalBP sumL)

sumL :: forall n m z.
        (KnownNat n,
         KnownNat m,
         Reifies z W)
     => BVar z (L m n)
     -> BVar z (R m)
sumL = liftOp1 . op1 $ \(L (Dim (Dim m))) -> (mkR $ A.fold (+) 0 m,
                                              (evalBP konstL))
sumAllL :: forall n m z.
        (KnownNat n,
         KnownNat m,
         Reifies z W)
     => BVar z (L m n)
     -> BVar z (R 1)
sumAllL = liftOp1 . op1 $ \(L (Dim (Dim m))) -> (pack $ A.flatten $ A.foldAll (+) 0 m,
                                                 (evalBP konstL) . (evalBP konstR))
-- WHAT A MONSTER. Could be HEAVILY simiplified, but w/e.
maximumR :: forall n z. (KnownNat n, Reifies z W) => BVar z (R n) -> BVar z (R 1)
maximumR = liftOp1 . op1 $ \x -> let (A.T2 (A.unindex1 -> i) max') =   
                                                    A.the . 
                                                      A.fold1 (\((A.T2 idx m))
                                                         (A.T2 idx' m') ->
                                                          A.cond (m A.>m') (A.T2 idx m) (A.T2 idx' m')
                                                       )
                                                    . A.indexed
                                                    . unpack $ x
                                     n = fromInteger $ natVal (Proxy @n)
                                 in (pack . A.flatten $ A.unit max',
                                     const .
                                     pack $
                                     A.generate (A.constant $ Z:.n) (\(A.unindex1 -> idx) ->
                                                                       A.cond (idx A.== i) 1 0)
                                    )
                                   
maximumL :: forall m n z. (KnownNat n, KnownNat m, Reifies z W) => BVar z (L m n) -> BVar z (R m)
maximumL = liftOp1 . op1 $ \x -> let (idxs,maxs) =  A.unzip .
                                                    A.fold1 (\(A.T2 idx m) (A.T2 idx' m') ->
                                                               A.cond (m A.> m') (A.T2 idx m) (A.T2 idx' m'))
                                                    . A.indexed
                                                    . unpack
                                                    $ x
                                     n = fromInteger $ natVal (Proxy @n) :: Int
                                     grad d = pack $
                                              A.imap (\idx (A.T2 idx' val) ->
                                                        A.cond ((A.unindex2 idx) A.== (A.unindex2 idx')) val 0.0)
                                              $ A.replicate (A.constant $ Z:. A.All :. n)
                                              $ A.zip idxs (unpack d) 
                                 in (pack maxs,
                                     \d -> grad d)
                                     
softMaxR :: (Reifies z W, KnownNat n) => BVar z (R n) -> BVar z (R n)
softMaxR x = expx / (konstR $ sumR expx) 
  where
    expx = exp x

transpose :: (Reifies z W, KnownNat m, KnownNat n)
          => BVar z (L m n)
          -> BVar z (L n m)
transpose = isoVar (pack . A.transpose . unpack) (pack . A.transpose . unpack)

softMaxL :: forall n m z.
            (Reifies z W,
             KnownNat n,
             KnownNat m)
         => BVar z (L m n)
         -> BVar z (L m n)
softMaxL x = expx / (konstL $ sumL expx) 
  where
    expx = exp x
linReg :: (KnownNat n, KnownNat m, Reifies z W)
       => BVar z ((L m n) ⊗ (R m))
       -> BVar z (R n)
       -> BVar z (R m)
linReg (w:*:b) v = w #> v + b

squaredErrorR :: forall n z. (KnownNat n, Reifies z W)
             => BVar z (R n) -> BVar z (R n) -> BVar z (R 1) 
squaredErrorR x y = (1/n)*(sumR (x-y)**2)
  where n = fromInteger $ natVal (Proxy @n) 

crossEntropyR :: (KnownNat n, Reifies z W)
             => BVar z (R n) -> BVar z (R n) -> BVar z (R 1) 
crossEntropyR x y = negate . sumR $ y * (log x)
clipBy :: Reifies z W
       => BVar z (Exp ℝ)
       -> BVar z (Exp ℝ)
       -> BVar z (Exp ℝ)
clipBy epsilon x = cmpr (A.<) (cmpr (A.>) x epsilon) (1-epsilon)
crossEntropyL :: forall m n z. (KnownNat m, KnownNat n, Reifies z W)
              => BVar z (L m n) -> BVar z (L m n) -> BVar z (R 1) 
              -- => Error (L m n) (R 1)
crossEntropyL x y = (/ m) . negate . sumAllL $ y * (log $ mapB (clipBy 1e-7) x)
  where m = fromInteger $ natVal (Proxy @m)

squaredErrorL :: forall m n z. (KnownNat n, KnownNat m, Reifies z W)
             => BVar z (L m n)
             -> BVar z (L m n)
             -> BVar z (R 1)
squaredErrorL x y = sumR $ 1/n*(sumL ((x-y)*(x-y)))
  where n = fromInteger $ natVal (Proxy @n) 
        m = fromInteger $ natVal (Proxy @m)


fullyConnected :: (KnownNat n, KnownNat m, Reifies z W)
               => (BVar z (R m) -> BVar z (R m)) 
               -> BVar z ((L m n) ⊗ (R m))
               -> BVar z (R n)
               -> BVar z (R m)
fullyConnected act (w :*: b) v = act $ (w #> v)+b

fullyConnectedL :: (KnownNat n, KnownNat m, KnownNat k, Reifies z W)
               => (BVar z (L m k) -> BVar z (L m k)) 
               -> BVar z ((L n k) ⊗ (L m k))
               -> BVar z (L m n)
               -> BVar z (L m k)
fullyConnectedL act (w :*: b) v = act $ (v <> w)+b
----------------------------------------------------------------------------------



src/Lift.hs
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE Rank2Types #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE ViewPatterns #-}
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE TypeApplications #-}

module Lift where
import GHC.TypeLits
import qualified Data.Array.Accelerate as A
import Data.Array.Accelerate ( Z(..), (:.)(..))
import Control.Category
import Control.DeepSeq
import Prelude hiding ((.))
import Numeric.Backprop
import Data.Proxy (Proxy(..))
import Data.Array.Accelerate.Numeric.LinearAlgebra hiding (Vector, Matrix)
import GHC.Generics (Generic)
import Data.Binary
--------------------------------------------------------------------------------
type ℝ = Float 
type Vector a = A.Acc (A.Vector a)
type Matrix a = A.Acc (A.Matrix a)

newtype Dim (n :: Nat) t = Dim t
  deriving (Show, Generic, Binary, NFData)

lift1F
  :: (c t -> c t)
  -> Dim n (c t) -> Dim n (c t)
lift1F f (Dim v) = Dim (f v)

lift2F
  :: (c t -> c t -> c t)
  -> Dim n (c t) -> Dim n (c t) -> Dim n (c t)
lift2F f (Dim u) (Dim v) = Dim (f u v)
--------------------------------------------------------------------------------

newtype R n = R (Dim n (Vector ℝ))
  deriving (Generic, Num, Fractional, Floating)
newtype L m n = L (Dim m (Dim n (Matrix ℝ)))
  deriving (Generic)

mkL :: Matrix ℝ -> L m n
mkL x = L (Dim (Dim x))

mkR :: Vector ℝ -> R n
mkR = R . Dim
--------------------------------------------------------------------------------

type V n t = Dim n (Vector t)

ud :: Dim n (Vector t) -> Vector t
ud (Dim v) = v

mkV :: forall (n :: Nat) t . t -> Dim n t
mkV = Dim

instance (KnownNat n, Num (A.Exp t), Numeric t) => Num (Dim n (Vector t)) where
  (+) = lift2F $ A.zipWith (A.+)
  (*) = lift2F $ A.zipWith (A.*)
  (-) = lift2F $ A.zipWith (A.-)
  abs = lift1F $ A.map abs
  signum = lift1F $ A.map signum
  negate = lift1F $ A.map (A.* (-1))
  fromInteger = Dim . A.fill (A.constant $ Z:.n) . A.fromInteger 
    where n = fromInteger $ natVal (Proxy @n) :: Int

instance (KnownNat n, Num (A.Exp t), Fractional (A.Exp t), Numeric t) => Fractional (Dim n (Vector t))
  where
    fromRational = Dim . A.fill (A.constant $ Z:.n) . A.fromRational
      where n = fromInteger $ natVal (Proxy @n) :: Int
    (/) = lift2F $ A.zipWith (A./)
instance (KnownNat n, Fractional (A.Exp t), Floating (A.Exp t), Numeric t) => Floating (Dim n (Vector t)) where
    sin   = lift1F $ A.map sin
    cos   = lift1F $ A.map cos
    tan   = lift1F $ A.map tan
    asin  = lift1F $ A.map asin
    acos  = lift1F $ A.map acos
    atan  = lift1F $ A.map atan
    sinh  = lift1F $ A.map sinh
    cosh  = lift1F $ A.map cosh
    tanh  = lift1F $ A.map tanh
    asinh = lift1F $ A.map asinh
    acosh = lift1F $ A.map acosh
    atanh = lift1F $ A.map atanh
    exp   = lift1F $ A.map exp
    log   = lift1F $ A.map log
    sqrt  = lift1F $ A.map sqrt
    (**)  = lift2F $ A.zipWith (**)
    pi = Dim . A.fill (A.constant $ Z:.n) $ A.pi
      where n = fromInteger $ natVal (Proxy @n) :: Int

--------------------------------------------------------------------------------
instance (KnownNat n, KnownNat m, Num (A.Exp t), Numeric t) => Num (Dim m (Dim n (Matrix t)))
  where
    (+) = (lift2F . lift2F) $ A.zipWith (+)
    (*) = (lift2F . lift2F) $ A.zipWith (*)
    (-) = (lift2F . lift2F) $ A.zipWith (-)
    abs = (lift1F . lift1F) $ A.map abs
    signum = (lift1F . lift1F) $ A.map signum
    negate = (lift1F . lift1F) $ A.map negate
    fromInteger = Dim . Dim . A.fill (A.constant $ Z:.m:.n) . A.fromInteger
      where n = fromInteger $ natVal (Proxy @n) :: Int
            m = fromInteger $ natVal (Proxy @m) :: Int


instance (KnownNat n, KnownNat m, Fractional (A.Exp t), Numeric t) => Fractional (Dim m (Dim n (Matrix t)))
  where
    fromRational = Dim . Dim . A.fill (A.constant $ Z:.m:.n) . A.fromRational
      where n = fromInteger $ natVal (Proxy @n) :: Int
            m = fromInteger $ natVal (Proxy @m) :: Int
    (/) = (lift2F.lift2F) $ A.zipWith (/)

instance (KnownNat n, KnownNat m, Fractional (A.Exp t), Floating (A.Exp t), Numeric t) => Floating (Dim m (Dim n (Matrix t))) where
    sin   = (lift1F . lift1F) $ A.map sin
    cos   = (lift1F . lift1F) $ A.map cos
    tan   = (lift1F . lift1F) $ A.map tan
    asin  = (lift1F . lift1F) $ A.map asin
    acos  = (lift1F . lift1F) $ A.map acos
    atan  = (lift1F . lift1F) $ A.map atan
    sinh  = (lift1F . lift1F) $ A.map sinh
    cosh  = (lift1F . lift1F) $ A.map cosh
    tanh  = (lift1F . lift1F) $ A.map tanh
    asinh = (lift1F . lift1F) $ A.map asinh
    acosh = (lift1F . lift1F) $ A.map acosh
    atanh = (lift1F . lift1F) $ A.map atanh
    exp   = (lift1F . lift1F) $ A.map exp
    log   = (lift1F . lift1F) $ A.map log
    sqrt  = (lift1F . lift1F) $ A.map sqrt
    (**)  = (lift2F . lift2F) $ A.zipWith (**)
    pi    = Dim . Dim . A.fill (A.constant $ Z:.m:.n) $ pi
      where n = fromInteger $ natVal (Proxy @n) :: Int
            m = fromInteger $ natVal (Proxy @m) :: Int
 
--------------------------------------------------------------------------------
lift1L f (L v) = L (f v)
lift2L f (L a) (L b) = L (f a b)
instance (KnownNat n, KnownNat m) =>  Num (L n m)
  where
    (+) = lift2L (+)
    (*) = lift2L (*)
    (-) = lift2L (-)
    abs = lift1L abs
    signum = lift1L signum
    negate = lift1L negate
    fromInteger = L . fromInteger

instance (KnownNat n, KnownNat m) => Fractional (L n m)
  where
    fromRational =  L . fromRational 
    (/) = lift2L (/)

instance (KnownNat n, KnownNat m) => Floating (L n m) where
    sin   = lift1L sin
    cos   = lift1L cos
    tan   = lift1L tan
    asin  = lift1L asin
    acos  = lift1L acos
    atan  = lift1L atan
    sinh  = lift1L sinh
    cosh  = lift1L cosh
    tanh  = lift1L tanh
    asinh = lift1L asinh
    acosh = lift1L acosh
    atanh = lift1L atanh
    exp   = lift1L exp
    log   = lift1L log
    sqrt  = lift1L sqrt
    (**)  = lift2L (**)
    pi    = L pi


instance (KnownNat n) => Show (R n) where
  show (R (Dim a)) =   "ℝ " ++ show (natVal (undefined :: Proxy n)) ++ " " ++ show a


tensorflow/tensorflowTest.py
import tensorflow as tf
mnist = tf.keras.datasets.mnist
bsize = 128
epchs = 10
target = "GPU"

(x_train, y_train),(x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0
y_train, y_test = tf.keras.utils.to_categorical(y_train, 10), tf.keras.utils.to_categorical(y_test, 10)
y_train.shape, y_test.shape

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(512, activation=tf.nn.relu,
kernel_initializer='RandomUniform', bias_initializer='RandomUniform'),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax,
kernel_initializer='RandomUniform', bias_initializer='RandomUniform')
])
model.compile(optimizer='sgd',
              loss='categorical_crossentropy',
              metrics=[tf.keras.metrics.categorical_accuracy])

hist = model.fit(x_train, y_train, epochs=epchs, batch_size=bsize, shuffle=False)
loss = hist.history['loss']
acc = hist.history['categorical_accuracy']
lossData = open("tf{}lossb{}e{}.data".format(target,bsize,epchs), "w")
accData = open("tf{}accb{}e{}.data".format(target,bsize,epchs), "w")
for i in range(len(loss)):
  lossData.write("{}.0 {}\n".format(i+1,loss[i]))
  accData.write("{}.0 {}\n".format(i+1,acc[i]))


src/MNIST.hs
{-# LANGUAGE ViewPatterns #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}

module MNIST where
import Ops 
import Lift 
import Backprop as AB
import Data.IDX
import qualified Data.Array.Accelerate as A
import Data.Array.Accelerate (Z(..),(:.)(..) )
import qualified Data.Vector.Unboxed as VU
import Prelude as P
import Control.Monad (forM_, forM)
import Control.Monad.Trans.Maybe (MaybeT(..))
import Control.Monad.Trans.State.Lazy as S
import Control.Monad.IO.Class
import Control.Arrow ((***))
import Data.Tuple (swap)
import Data.Proxy 
import Numeric.Backprop
import GHC.TypeLits
import System.Directory
import Data.Binary
import System.IO

-- Temporary chunksof code
build :: ((a -> [a] -> [a]) -> [a] -> [a]) -> [a]
build g = g (:) []

chunksOf :: Int -> [e] -> [[e]]
chunksOf i ls = P.map (take i) (build (splitter ls)) where
  splitter :: [e] -> ([e] -> a -> a) -> a -> a
  splitter [] _ n = n
  splitter l c n  = l `c` splitter (drop i l) c n
-- 
-- Stolen and modified from JLE
loadMNISTB  :: forall m. KnownNat m
            => FilePath
            -> FilePath
            -> IO (Maybe [(SL m 784, SL m 10)])
loadMNISTB fpI fpL = runMaybeT $ do i <- MaybeT          $ decodeIDXFile       fpI
                                    l <- MaybeT          $ decodeIDXLabelsFile fpL
                                    d <- MaybeT . return $ labeledIntData l i
                                    return $ P.map (swap . (mkLabel *** mkImage) . P.unzip) . P.init $ chunksOf m d
                                      where 
                                        m = fromInteger $ natVal (Proxy @m) :: Int
                                        mkImage :: [VU.Vector Int] -> (SL m 784)
                                        mkImage =  pack . A.fromList (Z:.m:.784) . VU.toList . VU.map (\i -> fromIntegral i / 255) . VU.concat
                                        mkLabel :: [Int] -> (SL m 10)
                                        mkLabel =  pack . A.fromList (Z:.m:.10) . P.concat . P.map (\n -> P.map (\x -> if (x==n) then 1.0 else 0.0) [0..9])
-- 



writeMNIST :: forall m. (KnownNat m) => IO(Maybe [(SL m 784, SL m 10)])
writeMNIST = do let m = natVal (Proxy @m)
                    path = "mnist" P.++ "b" P.++ (show m) P.++ ".hldata"
                bool <- doesFileExist path
                if(bool) then decodeFile path >>= return
                  else
                  do
                  mnist <- loadMNISTB "data/train-images-idx3-ubyte" "data/train-labels-idx1-ubyte"
                  encodeFile path mnist
                  return mnist


type Accuracy = Float
type Loss = Float
type Dataset a b = [(a, b)]
type Epochs = Int
data TrainState p a b = TrainState (TrainConfig p a b) !Accuracy !Loss !p 
data TrainConfig p a b = TrainConfig Epochs (Dataset a b) (p -> (a,b) -> (SR 2, p))


sgd :: (Parameter p,
        Static a,
        Static b)
    =>  Rate
    -> (Learn p a b)
    -> (Error b (R 1))
    -> (b -> b -> R 1) 
    -> (Run p -> (Run a, Run b) -> (SR 2, Run p))
sgd rate learn err metric  = curry $ run1 sgdAcc 
  where sgdAcc (p:::(a:::b)) = 
          let newParams = update rate p grad 
              grad = errorGrad learn err a b p
              eval = evalBP2 learn newParams a
              loss = evalBP2 err eval b 
              acc = metric eval b
              lossacc = evalBP2 (AB.++) acc loss
          in lossacc:::newParams



xtract :: SR 2 -> (Accuracy, Accuracy)
xtract =  (\[a,b] -> (a,b)) . A.toList . unpack
trainOverBatches :: forall h m n k.
                    (KnownNat h,
                     KnownNat m,
                     KnownNat n,
                     KnownNat k)
                 => StateT (TrainState ((SL n h, SL m h), (SL h k, SL m k))
                             (SL m n) (SL m k))
                    IO ()
trainOverBatches  =
  do
    (TrainState (TrainConfig epochs dataset opt) _ _ _) <- S.get
    let target = "GPU"
        bsize = natVal (Proxy @m)
        prefix = "hl" P.++ target
        suffix = "b" P.++ (show bsize) P.++ "e" P.++ (show epochs) P.++ ".data"
    accFile <- liftIO $ openFile (prefix P.++ "acc" P.++ suffix) WriteMode
    lossFile <- liftIO $ openFile (prefix P.++ "loss" P.++ suffix) WriteMode
    forM_ [1..epochs] $ \e ->
      -- Iterate over epochs
      do
        liftIO $ putStrLn $ "Epoch " P.++ (show e)
        S.modify (\(TrainState tc _ _ p) -> TrainState tc 0 0 p)
        -- One pass over training set 
        epochHistory <- forM (P.zip [1..] dataset) $ \(b,chnk) ->
          do
            liftIO $ putStrLn $ "Batch " P.++ (show b)
            (TrainState tc acc loss p) <- S.get
            let (lossAcc,p') = opt p chnk
                (accInc,lossInc) = xtract lossAcc
                (acc',loss') = if (b==1) then (accInc,lossInc) else (acc+accInc, loss+lossInc)
                (scA,scL) = (acc'/(P.fromIntegral b),loss'/(P.fromIntegral b))
            liftIO $ print (scA, scL)
            S.put (TrainState tc acc' loss' p')
            return (scA, scL)
        let epochPrefix val = (show e) P.++ " " P.++ (show val)
            (epochAcc, epochLoss) =  P.last epochHistory
        liftIO $ do hPutStrLn accFile (epochPrefix epochAcc) 
                    hPutStrLn lossFile (epochPrefix epochLoss)
    liftIO $ do hClose accFile
                hClose lossFile

trainMNIST = do
  Just mnistdb <- liftIO $ writeMNIST @128 -- Batch size
  let tc = TrainConfig epochs mnistdb mnistOpt
      epochs = 10
      emptyState = TrainState tc 0 0
  initialWeights <- sample >>= (return . run . AB.map (\x -> (x-0.5)/10) . use)
  hSetBuffering stdout NoBuffering
  evalStateT trainOverBatches (emptyState initialWeights)
  return ()
  where
    mnistOpt = sgd 0.01 (fullyConnectedL (mapB relu) ~> fullyConnectedL @512 softMaxL) crossEntropyL computeAccuracy



computeAccuracy :: forall m n. (KnownNat m, KnownNat n) => L m n -> L m n -> (R 1)
computeAccuracy eval b  = let accInc =   (/ m)
                                          . pack
                                          . A.flatten
                                          . A.fold1 (+)
                                          . A.map (\(A.T2 (A.T2 idx1 _) (A.T2 idx2 _)) -> A.cond ((A.unindex2 idx1) A.== (A.unindex2 idx2)) 1 0)  
                                          . A.fold1 (\(A.T2 idxs1 idxs2) (A.T2 idxs1' idxs2') ->
                                                        let compareIdxs (A.T2 idx x) (A.T2 idx' y) = A.cond (x A.> y) (A.T2 idx x) (A.T2 idx' y)
                                                            newIdx1 = compareIdxs idxs1 idxs1'
                                                            newIdx2 = compareIdxs idxs2 idxs2'
                                                        in (A.T2 newIdx1 newIdx2))
                                          $ A.zip (A.indexed $ unpack eval) (A.indexed $ unpack b) :: R 1
                              m = fromIntegral $ natVal (Proxy @m)
                          in (accInc)
 
--main :: IO ()
-- main =  trainMNIST 
-- trainMNIST = do
--     -- Just test <- liftIO $ loadMNISTB "data/t10k-images-idx3-ubyte" "data/t10k-labels-idx1-ubyte" 
--     -- let testAccuracy w = (\(a,b) -> let len = P.fromIntegral $ P.length test in (a/len,b/len)) $
--     --                        foldl' (\(a,b) (c,d)
--     --                                 -> (a+c,b+d)) (0,0) $
--     --                        (P.map (xtract  .
--     --                                (curry .
--     --                                 run1 $ \(p:::(a:::b)) ->
--     --                                          let eval = evalBP2 learn p a
--     --                                              loss = evalBP2 err eval b
--     --                                              acc = computeAccuracy ((evalBP2 learn) p a) b
--     --                                          in (evalBP2 (AB.++) acc loss)
--     --                                ) w
--     --                               )
--     --                        )
--     --                        test
--     trainOverBatches  
--     --(TrainState _ finalAcc finalLoss _) <- S.get 
--     -- liftIO $ print $ testAccuracy finalWeights
--     -- liftIO $ writeHistory history
--     return ()
